rocketpool-cli/update/update.go

```diff
@@ -0,0 +1,222 @@
+		if err != nil {
+			return fmt.Errorf("could not parse version number from release name '%s': %w", name, err)
+		}
+	} else {
+		return fmt.Errorf("unexpected Github API response format")
+	}
+
+	// Check this version against the currently installed version
+	if !c.Bool("force") {
+		currentVersion, err := semver.Make(shared.RocketPoolVersion)
+		if err != nil {
+			return fmt.Errorf("could not parse local Rocket Pool version number '%s': %w", shared.RocketPoolVersion, err)
+		}
+		switch latestVersion.Compare(currentVersion) {
+		case 1:
+			fmt.Printf("Newer version avilable online (v%s). Downloading...\n", latestVersion.String())
+		case 0:
+			fmt.Printf("Already on latest version (v%s). Aborting update\n", latestVersion.String())
+			return nil
+		default:
+			fmt.Printf("Online version (v%s) is lower than running version (v%s). Aborting update\n", latestVersion.String(), currentVersion.String())
+			return nil
+		}
+	} else {
+		fmt.Printf("Forced update to v%s. Downloading...\n", latestVersion.String())
+	}
+
+	// Download the new binary to same folder as the running RP binary, as `rocketpool-vX.X.X`
+	var ClientURL = fmt.Sprintf(ReleaseBinaryURL, latestVersion.String(), runtime.GOOS, runtime.GOARCH)
+	resp, err = http.Get(ClientURL)
+	if err != nil {
+		return fmt.Errorf("error while downloading %s: %w", ClientURL, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("request failed with code %d", resp.StatusCode)
+	}
+
+	ex, err := os.Executable()
+	if err != nil {
+		return fmt.Errorf("error while determining running rocketpool location: %w", err)
+	}
+	var rpBinDir = filepath.Dir(ex)
+	var fileName = filepath.Join(rpBinDir, "rocketpool-v"+latestVersion.String())
+	output, err := os.OpenFile(fileName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+	if err != nil {
+		return fmt.Errorf("error while creating %s: %w", fileName, err)
+	}
+	defer output.Close()
+
+	_, err = io.Copy(output, resp.Body)
+	if err != nil {
+		return fmt.Errorf("error while downloading %s: %w", ClientURL, err)
+	}
+
+	// Verify the signature of the downloaded binary
+	if !c.Bool("skip-signature-verification") {
+		var pubkeyUrl = fmt.Sprintf(SigningKeyURL, latestVersion.String())
+		output.Seek(0, io.SeekStart)
+		err = checkSignature(ClientURL+".sig", pubkeyUrl, output)
+		if err != nil {
+			return fmt.Errorf("error while verifying GPG signature: %w", err)
+		}
+	}
+
+	// Prompt for confirmation
+	if !(c.Bool("yes") || cliutils.Confirm("Are you sure you want to update? Current Rocketpool Client will be replaced.")) {
+		fmt.Println("Cancelled.")
+		return nil
+	}
+
+	// Do the switcheroo - move `rocketpool-vX.X.X` to the location of the current Rocketpool Client
+	err = os.Remove(ex)
+	if err != nil {
+		return fmt.Errorf("error while removing old rocketpool binary: %w", err)
+	}
+	err = os.Rename(fileName, ex)
+	if err != nil {
+		return fmt.Errorf("error while writing new rocketpool binary: %w", err)
+	}
```

Comment:
When replacing the binary, the key is the the inode is different to the original executable. Since we have a new file created during the download process. So this is pretty safe. If we were using something else like truncate/copy, then we might still have the original inode, which can be problematic for the OS to deal with.